File: 621.task-scheduler.js
/*
 * @lc app=leetcode id=621 lang=javascript
 *
 * [621] Task Scheduler
 */

// @lc code=start
/**
 * @param {string[]} tasks
 * @param {number} n
 * @return {number}
 */
const leastInterval = function (tasks, n) 
{
    /**
     * 任务与其要被执行的次数
     * @type {Map<string, number>}
     */
    const taskToExecutionCount = new Map();
    let count = 0;
    for (const task of tasks)
    {
        count = taskToExecutionCount.get(task) ?? 0;
        taskToExecutionCount.set(task, count + 1);
    }
    const taskToCountEntries = Array.from(taskToExecutionCount.entries()).sort((a, b) => b[1] - a[1]);
    /** 最多的那种任务的执行次数 */
    const biggestExecutedCount = taskToCountEntries[0][1];
    /** 有多少种任务的执行次数都是最多的 */
    let tasksWithBiggestExecutedCountAmount = 1;

    for (let i = 1; i < taskToCountEntries.length; i++)
    {
        if (taskToCountEntries[i][1] === biggestExecutedCount)
        {
            tasksWithBiggestExecutedCountAmount++;
        }
        else
        {
            break;
        }
    }
    /** 除了个数最多的任务之外，剩余任务的数量 */
    const leftTasksAmount = tasks.length - tasksWithBiggestExecutedCountAmount * biggestExecutedCount;
    const slotsAmountPerGroup = n + 1;
    const idleSlotsAmount = (slotsAmountPerGroup - tasksWithBiggestExecutedCountAmount) * (biggestExecutedCount - 1);

    if (idleSlotsAmount >= leftTasksAmount)
    {
        return slotsAmountPerGroup * (biggestExecutedCount - 1) + tasksWithBiggestExecutedCountAmount;
    }
    else
    {
        return tasks.length;

    }
};
// @lc code=end

File: Daily/621.task-scheduler.cpp
/*
 * @lc app=leetcode id=621 lang=cpp
 *
 * [621] Task Scheduler
 */
#include <algorithm>
#include <map>
#include <vector>

using std::map;
using std::vector;

// @lc code=start
class Solution {
 public:
  int leastInterval(vector<char>& tasks, int n) {
    const int TASK_COUNT = tasks.size();

    vector<int> taskFrequency(26, 0);
    int maxTaskFrequency = 0;
    for (char task : tasks) {
      taskFrequency[task - 'A']++;
      maxTaskFrequency = std::max(maxTaskFrequency, taskFrequency[task - 'A']);
    }

    const int taskWithMaxFrequencyCount =
        std::count_if(taskFrequency.cbegin(), taskFrequency.cend(),
                      [maxTaskFrequency](auto element) {
                        return element == maxTaskFrequency;
                      });

    return std::max(TASK_COUNT, (maxTaskFrequency - 1) * (n + 1) +
                                    taskWithMaxFrequencyCount);
  }
};
// @lc code=end

int main() {
  Solution sol;
  vector<char> tasks = {'A', 'A', 'A', 'B', 'B', 'B'};
  sol.leastInterval(tasks, 2);
}